### DISCUSSION TITLE
arrtwork retreiving improvement

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#48] [Ideas]
Yesterday I had internet breakup from my ISP.

My kodi system was functioning Ok (almost everything is on my NAS) and kodi_panel worked partially.
All layouts that use artwork retrieved from the internet were not working (stuck on "Waiting to connect with Kodi...")

As soon As i changed media to use a layout that did not need internet kodi_panel was working Ok.

I mainly tested Tvshow artwork (maybe audio artwork would have a different behavior?)

I believe it is because Kodi_panel could not retreive the artwork from internet. 
However as Kodi and/or libreelec cache all artwork kodi was able to display it.

Here is a few ideas to improve this behavior:

1. Retrieve artwork directly from kodi (and not form the same adress) 
As kodi cache the media. it would allow a fully functional Kodi_panel even without internet or when the artwork become unavailable from internet (and also speed up the process for slow connection) 

2. Have a timeout feature: when artwork can not get retrieved from internet, display the layout without artwork (or default picture instead)

**Comment by mattblovell** [comment]
> Retrieve artwork directly from kodi (and not form the same adress)

kodi_panel only retrieves artwork from the URL that Kodi itself provides.  It is certainly possible that Kodi directs image retrieval to someplace else, as I'll discuss momentarily, but there's not much that kodi_panel can do about that.  Kodi is the central source for all information, after all.

In my (music only) setup, all music files are hosted on my PC for DLNA playback through Kodi (using Kodi as the renderer).  The URL that Kodi provides during playback points back to the PC.  If kodi_panel elected _not_ to use that URL (as it no longer references the kodi web server), what else would it use?

> Have a timeout feature: when artwork can not get retrieved from internet, display the layout without artwork (or default picture instead)

What version of kodi_panel are you using?  

The "Waiting to connect to Kodi" message is indicative of _local_ communication woes, unable to get a JSON-RPC "ping" to work.  Code in that area was made more robust 5 to 6 months ago.

I recently had to completely re-install CoreELEC (the microSD card I was using had failed).  The same running instance of kodi_panel was able to survive multiple disruptions in communication, as I was powering the Odroid C4 on and off and rebooting Kodi.  It is difficult to exercise all of those code paths, but I think most scenarios are working.  If your installation is not the most recent, though, we'd have to start with an update before any further diagnosis.

  Matt
